ReadMe.md
## HOME AUTOMATION README
---
### How To:
---
**For:** Arduino Atmega2560

**Flashing to Arduino:** 

---

### Home.ino
##### Void Setup:
The void setup is where all the setup functions are done.

1. The connector function also handles the messages recieved and sent.
2. The connector function will be explained in more detail below. 
```c++
void setup() {
//Call connector function located in Connection.h
connector();
//Call pinSetup function located in PinLogic.h
pinSetup();
}
``` 
##### Void Loop:
The void loop is used to check connection status and reconnect function.
```c++
void loop(){
//Call checkConnect function located in Connection.h
checkConnect();
}
``` 
### PinLogic.h
##### Void pinSetup:
Inside pinSetup the variables are declared and pins for the arduino are setup.

```c++
void pinSetup() {
    //Variables set to instantiate pin values 
    int pin = 0;
    //Delay set to allow the previous function in Home.ino to finish.
    delay(1500);
    //Open serial listner on 9600 baud rate.
    Serial.begin(9600);
    //For loop declared to loop through the pin values to be set.
    for (int pin = 22; pin < 43; pin++) {
        //Set the pin values to output.
        pinMode(pin, OUTPUT);
        //Check state of each pin initially.
        digitalRead(pin);
        //Set each pin to preffered state. 
        digitalWrite(pin, HIGH);
    }
}
``` 
### Connection.h
##### Globals:


```c++
#include "log.h"
//Client mac address specified.
byte mac[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
// Setup network address for Client & Server.
IPAddress ip(192, 168, 0, 0);
IPAddress server(192, 168, 0, 0);
//Declare Ethernet & MQTT client libraries.
EthernetClient ethClient;
PubSubClient mqttClient(ethClient);
```
##### Void callback:

Detailed comments.
```c++
//Function callback handles incoming & outgoing messages.
void callback(char* topic, byte* payload, unsigned int length) {
  //Declaring variables for use.
  String content = "";
  char character;
  //For loop is used to read the payload in bytes.
  for (int num = 0; num < length; num++) {
    //The payload is converted into char variable.
    character = payload[num];
    //Here the variable character is converted from char to String.
    content.concat(character);
  }
  //Debug messages are logged to serial.
  logger("callback_MQTT_topic", "Topic:" + String(topic));
  logger("callback_MQTT_payload", "Payload:" + content);
  //The content variable is convert to an integer.
  int pinval = content.toInt();
  //Debug messages are logged to serial.
  logger("callback_set_pinval", "Pin Value:" + pinval);
  //Declare new string variable use.
  String pinvalstr;
  //Condition to cancel out pins that are not set.
  if (pinval > 21 && pinval < 43) {
    //Read state of pins.
    int swt = digitalRead(pinval);
    //Debug messages are logged to serial.
    logger("callback_set_pinstate", "Pin State:" + swt);
    //Condition checked and switch state set.
    swt = swt > 0 ? LOW : HIGH;
    //Condition sets pinvalstr to a unique message to return on or off value.
    pinvalstr = swt == LOW ? content + "0" : content + "1" ;
    //Integer declared to count pinvalstr length.
    int str_len = pinvalstr.length() + 1;
    //Declaring char array.
    char char_array[str_len];
    //Convert pinvalstr into char array.
    pinvalstr.toCharArray(char_array, str_len);
    //Condition to check if it is connected to server.
    if (mqttClient.connect("anorakClient")) {
      //Send message to MQTT server.
      mqttClient.publish("Homeret", char_array);
    }
    //Set the pin number to the specific state determined above.
    digitalWrite(pinval, swt);
    //Debug messages are logged to serial.
    logger("callback_final_check", "Pin Value:" + String(pinval) + "Pin State:" + swt);
  }
}
```
##### Void connector:
```c++
//Function connector sets up the connection to the server.
void connector() {
    //Server address and port specified.
    mqttClient.setServer(server, 1883);
    //Call library function to use callback.
    mqttClient.setCallback(callback);
    //Start ethernet client connection to connect to server. 
    Ethernet.begin(mac, ip);
}
```
##### Void reconnect:
```c++
//Function reconnect will check state of connection and reatempt if disconnected.
void reconnect() {
    // Loop until connected
    while (!mqttClient.connected()) {
        logger("reconnect", "Attempting MQTT connection.");
        // Attempt to reconnect.
        if (mqttClient.connect("anorakClient")) {
            logger("reconnect_success", "Connected!");
            // Once connected, publish a Message.
            mqttClient.publish("Home2", "connected");
            // Subscribe to specific topic.
            mqttClient.subscribe("Home2");
        }
        else {
            logger("reconnect_fail", "State:" + String(mqttClient.state()) + ", Reattempting connection in 5 seconds.");
            // Wait 5 seconds before retrying
            delay(5000);
        }
    }
    mqttClient.loop();
}
``` 



CHANGELOG.md
# Changelog
---
### V2.2:
        1.Removed checkConnect function and integrated 
          MQTT client loop into reconnect function within
          Connection.h.
        2.Home.ino code formatted.
        3.Add log.h library to print serial values for debug purposes.
        4.Added comment lines for Connection.h
        5.Included log.h as library in Connection.h & Home.ino
        6.Changed Mac & IP addresses to variable values inside Connection.h
        7.Added Connection.h content to ReadMe for better explination.
        8.Added Node-Red GUI source code as noderedGUI.json file.
---
**Added log.h**
```c++
void logger(String cat, String msgg) {
	Serial.println("[" + cat + "] " + msgg);
}
```
**Removed from Connection.h**

Reason: Not required,checkConnect was merged with reconnect function.

**V2.1 code:**
```c++
void connector() {
    mqttClient.setServer(server, 1883);
    mqttClient.setCallback(callback);
    Ethernet.begin(mac, ip);
}

void reconnect() {
    // Loop until connected
    while (!mqttClient.connected()) {
        Serial.print("Attempting MQTT connection...");
        // Attempt to connect
        if (mqttClient.connect("anorakClient")) {
            Serial.println("connected");
            // Once connected, publish a Message.
            mqttClient.publish("Home2", "connected");
            // Subscribe to specific Topic
            mqttClient.subscribe("Home2");
        }
        else {
            Serial.print("Failed to connect: ");
            Serial.print(mqttClient.state());
            Serial.println(" Will retry in 5 seconds.");
            // Wait 5 seconds before retrying
            delay(5000);
        }
    }
}

void checkConnect() {
    if (!mqttClient.connected()) {
        reconnect();
    }
    mqttClient.loop();
}
```
**TO**
**V2.2 code:**
```c++
//Function reconnect will check state of connection and reatempt if disconnected.
void reconnect() {
    // Loop until connected
    while (!mqttClient.connected()) {
        logger("reconnect", "Attempting MQTT connection.");
        // Attempt to reconnect.
        if (mqttClient.connect("anorakClient")) {
            logger("reconnect_success", "Connected!");
            // Once connected, publish a Message.
            mqttClient.publish("Home2", "connected");
            // Subscribe to specific topic.
            mqttClient.subscribe("Home2");
        }
        else {
            logger("reconnect_fail", "State:" + String(mqttClient.state()) + ", Reattempting connection in 5 seconds.");
            // Wait 5 seconds before retrying
            delay(5000);
        }
    }
    mqttClient.loop();
}
```
 
**Replaced in Home.ino**

Reason: Not needed, unecesary function.
        Replaced with reconnect function. 

**V2.1 code:**
```c++
void loop(){
    //Call checkConnect function located in Connection.h
    checkConnect();
}   
```
**To V2.2 code:**
```c++
void loop(){
    //Call reconnect function located in Connection.h
    reconnect();
}

```


---
### V2.1:
        1.Added information to ReadMe document about Home.ino
            i. Basic description of code 
        2.Added information to ReadMe document about PinLogic.h 
            i. Basic description of code.
        3.Added comment lines for Home.ino
        4.Added comment lines for PinLogic.h 
---
Removed from PinLogic.h

Reason: Not needed, code was not removed before V2.1. 
```c++
int spin = digitalRead(pin); 
int butval = 0;       
```
       


src/Connection.h
#include "log.h"
//Client mac address specified.
byte mac[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
// Setup network address for Client & Server.
IPAddress ip(192, 168, 0, 0);
IPAddress server(192, 168, 0, 0);
//Declare Ethernet & MQTT client libraries.
EthernetClient ethClient;
PubSubClient mqttClient(ethClient);
//Function callback handles incoming & outgoing messages.
//Function callback is explained in more detail in the ReadMe for more information.
void callback(char* topic, byte* payload, unsigned int length) {
    String content = "";
    char character;
    for (int num = 0; num < length; num++) {
        character = payload[num];
        content.concat(character);
    }
    logger("callback_MQTT_topic", "Topic:" + String(topic));
    logger("callback_MQTT_payload", "Payload:" + content);
    int pinval = content.toInt();
    logger("callback_set_pinval", "Pin Value:" + pinval);
    String pinvalstr;
    if (pinval > 21 && pinval < 43) {
        int swt = digitalRead(pinval);
        logger("callback_set_pinstate", "Pin State:" + swt);
        swt = swt > 0 ? LOW : HIGH;
        pinvalstr = swt == LOW ? content + "0" : content + "1";
        int str_len = pinvalstr.length() + 1;
        char char_array[str_len];
        pinvalstr.toCharArray(char_array, str_len);
        if (mqttClient.connect("anorakClient")) {
            mqttClient.publish("Homeret", char_array);
        }
        digitalWrite(pinval, swt);
        logger("callback_final_check", "Pin Value:" + String(pinval) + "Pin State:" + swt);
    }
}
//Function connector sets up the connection to the server.
void connector() {
    mqttClient.setServer(server, 1883);
    mqttClient.setCallback(callback);
    Ethernet.begin(mac, ip);
}
//Function reconnect will check state of connection and reatempt if disconnected.
void reconnect() {
    // Loop until connected
    while (!mqttClient.connected()) {
        logger("reconnect", "Attempting MQTT connection.");
        // Attempt to reconnect.
        if (mqttClient.connect("anorakClient")) {
            logger("reconnect_success", "Connected!");
            // Once connected, publish a Message.
            mqttClient.publish("Home2", "connected");
            // Subscribe to specific topic.
            mqttClient.subscribe("Home2");
        }
        else {
            logger("reconnect_fail", "State:" + String(mqttClient.state()) + ", Reattempting connection in 5 seconds.");
            // Wait 5 seconds before retrying
            delay(5000);
        }
    }
    mqttClient.loop();
}

